Fix: RandomShadow input type wrong

i-aki-y opened the pull request:
This PR ensures input image and mask array are contiguous in memory space to avoid incompatibility with cv:Mat data type used in OpenCV. (close #614 #1025)

As proposed in #1025, `np.ascontiguousarray` are inserted in the `Compose.__call__()`.
Since some transforms can output a non-contiguous array, `np.ascontiguousarray` is also applied after each transforms.

Note that the error is still raised if users apply a transform to a non-contiguous input individually (applying without Compose). In such a case, ensuring a contiguous input is a user's responsibility.

I think fixes for some base classes like`DualTransform` and `ImageOnlyTransform` (or `BasicTransform`) are needed to cover the individual transform case if needed.

I appreciate any comments and suggestions.

Dipet commented on albumentations/core/composition.py:
```diff
@@ -201,11 +201,13 @@ def __call__(self, *args, force_apply: bool = False, **data) -> typing.Dict[str,
 
             if check_each_transform:
                 data = self._check_data_post_transform(data)
+            data = self._make_targets_contiguous(data)
```
I think we do not need this check there because all transforms must return contiguous arrays. Better to write a test to check that each transform returns contiguous data.
Look at [test_image_only_augmentations_with_float_values](https://github.com/albumentations-team/albumentations/blob/master/tests/test_augmentations.py#L79) for an example

i-aki-y replied:
Thank you for your comment.
OK, I think I got how the test works.
I will examine the violating transforms.

Dipet commented on albumentations/core/composition.py:
```diff
@@ -271,6 +273,13 @@ def _check_args(self, **kwargs) -> None:
             if internal_data_name in check_bbox_param and self.processors.get("bboxes") is None:
                 raise ValueError("bbox_params must be specified for bbox transformations")
 
+    def _make_targets_contiguous(self, data):
+        targets = ["image", "mask"]
+        for target in targets:
+            if target in data and isinstance(data[target], np.ndarray) and not data[target].flags["C_CONTIGUOUS"]:
+                data[target] = np.ascontiguousarray(data[target])
+        return data
```
Thank you for your contribution.
I think it is better to write a generalized function that will covert all np.ndarrays to a contiguous format.
```python
@staticmethod
def _make_targets_contiguous(data: typing.Dict[str, typing.Any]) -> typing.Dict[str, typing.Any]:
    result = {}

    for key, value in data.items():
        if isinstance(value, np.ndarray):
            value = np.ascontiguousarray(value)
        result[key] = value

    return result
```

Also, we do not need this check `data[target].flags["C_CONTIGUOUS"]` because if the array is contiguous `np.ascontiguousarray` will return it unchanged and will not create a copy.

i-aki-y replied:
Yes. you're right! I will fix it.

i-aki-y commented:
I think I have made a test case to check all transforms. And I found that some transforms return non-contiguous output, the pytest summary is shown below.
Now I will fix these transforms to return contiguous results.

List of transforms that return non-contiguous output:
```
======================================================= short test summary info ========================================================
FAILED tests/test_augmentations.py::test_return_contiguous_ndarray[CenterCrop-params4] - assert False
FAILED tests/test_augmentations.py::test_return_contiguous_ndarray[ChannelShuffle-params6] - assert False
FAILED tests/test_augmentations.py::test_return_contiguous_ndarray[Crop-params9] - assert False
FAILED tests/test_augmentations.py::test_return_contiguous_ndarray[CropNonEmptyMaskIfExists-params11] - assert False
FAILED tests/test_augmentations.py::test_return_contiguous_ndarray[FDA-params16] - assert False
FAILED tests/test_augmentations.py::test_return_contiguous_ndarray[RandomCrop-params48] - assert False
FAILED tests/test_augmentations.py::test_return_contiguous_ndarray[RandomCropNearBBox-params49] - assert False
FAILED tests/test_augmentations.py::test_return_contiguous_ndarray[Transpose-params76] - assert False
```

i-aki-y commented:
I applied the result of the review comments.

I made a test to check all augmentation and fixed some of them to return contiguous ndarray.

I appreciate it if you review this one.

BloodAxe commented:
Hi! Thanks for your PR. This is indeed valuable improvement for the library. Yet i feel it may have impact on the performance.
Consider a following pipeline:
```
A.Compose([A.RandomCrop(), A.HorisontalFlip(), A.VerticalFlip()])
```
There are three consecutive augmentations that produce non-contiguous output. With the proposed change we will make 3 copies of the sliced data simply to make another slice. Which is clearly a waste of processing time. 

The initial reason why we want to have contiguous data is that OpenCV fails to handle image slices as input. With this knowledge in mind, I suggest to consider another solution:

1) Extend our unit tests to test all augmentations for non-contiguous inputs
2) For those, who fails, add a decorator function similar to `@preserve_shape` to ensure input for that function is contiguous:
```
def ensure_contiguous(func):

    @wraps(func)
    def wrapped_function(img, *args, **kwargs):
        img = np.require(img, requirements=np.C_CONTIGUOUS)
        result = func(img, *args, **kwargs)
        return result

    return wrapped_function
```
This change would ensure that we make a copy only when necessary and would be optimal from performance point of view.

3) As a final step - call a `_make_targets_contiguous` to ensure that outputs from the whole pipeline are also contiguous.

i-aki-y commented:
@BloodAxe Thank you for your suggestion. I will try your idea.

i-aki-y commented:
@BloodAxe @Dipet  I fixed the PR.

I modified the test case to detect transforms that raise an error with non-contiguous input instead of checking contiguous output. 
As a result of the test, I found that the `add_shasow` was the only function with a problem. 
So in this new approach, fixes became simpler than the previous version.

I appreciate it if you review this modification.

BloodAxe commented:
Thanks @i-aki-y, looking into it.

BloodAxe commented:
Thank your for your contribution. Merged)
